[Woo POS] Return to POS right after payment success

samiuelson opened the pull request:
## Return to POS right after payment success

Closes: #11628 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The intent of this PR is to further modify the existing IPP flow for the POS purpose. The goal is to exit the payment flow after payment capture success, before displaying the final success dialog – we want to show the custom "Payment Sucess" screen in POS instead.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Make a purchase in the POS mode. Observe that the payment success dialog is no longer visible, and the "green" Success Payment screen is shown automatically.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
* It's crucial to test the IPP flow against regression in Store Management mode. Existing IPP flow should remain unchanged—show dialog with success state and receipt handling options.
* Make a purchase in POS, and verify green "Payment success" screen is shown automatically after payment completes.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/70eaae7b-2af3-4132-bd83-5ef67701482a

- [x] I have considered adding unit tests for this change. If I decided not to add them, I have provided a brief explanation below (optional):
- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Return to POS right after payment success

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/229">19.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Improve code style

samiuelson committed:
Add unit tests

samiuelson committed:
Merge branch 'trunk' into 11628-return-to-pos-right-after-payment-success

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentViewModel.kt:
```diff
@@ -464,9 +464,19 @@ class CardReaderPaymentViewModel
+        }
+    }
+
+    private fun CardReaderFlowParam.PaymentOrRefund.isPOS(): Boolean {
```
np: I think semantically this more prop than a fun

```
val CardReaderFlowParam.PaymentOrRefund.isPOS: Boolean
    get() = this is CardReaderFlowParam.PaymentOrRefund.Payment &&
            this.paymentType == CardReaderFlowParam.PaymentOrRefund.Payment.PaymentType.WOO_POS
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentViewModel.kt:
```diff
@@ -464,9 +464,19 @@ class CardReaderPaymentViewModel
-        reFetchOrder()
+        if (arguments.paymentOrRefund.isPOS()) {
+            reFetchOrder()
+            onCancelPaymentFlow()
```
I am a bit concerned that we call on cancel flow here as it does some stuff that we may not need at all. E.g. tracking of the the flow cancelation, and show snack bar in some cases

Also, do we need to fetch the order for POS? If not, maybe then just:

```
if (arguments.paymentOrRefund.isPOS()) {
    triggerEvent(Exit) 
} else {...}
```
will do?

samiuelson replied:
> Also, do we need to fetch the order for POS?

Yes, it's necessary to fetch the order, otherwise, the flow will not return successful result to POS and Payment successful screen won't be shown.

> I am a bit concerned that we call on cancel flow here as it does some stuff that we may not need at all. E.g. tracking of the the flow cancelation, and show snack bar in some cases

Good point. I got rid of the redundant code here [58c1a9d](https://github.com/woocommerce/woocommerce-android/pull/12232/commits/58c1a9d6a2504a09aa3ea918e9b39c6ff6172513):
```kotlin
if (arguments.paymentOrRefund.isPOS) {
            launch {
                orderRepository.fetchOrderById(orderId)
                triggerEvent(Exit)
            }
        }
```
Let me know what you think.

kidinov replied:
> Yes, it's necessary to fetch the order, otherwise, the flow will not return successful result to POS and Payment successful screen won't be shown.

Oh, yeah, I see that we check the status of the order

We don't show any loading, but "capture payment" - not ideal, but I think it's ok for now

kidinov reviewed (commented):
I test both IPP and POS flow - works nicely

I left a couple of comments - please take a look

samiuelson committed:
Clean up the code

samiuelson commented:
> I test both IPP and POS flow - works nicely
> 
> I left a couple of comments - please take a look

Thanks for review @kidinov! I addressed the comments. Let me know what you think.

samiuelson committed:
Fix detekt's complaints

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentViewModel.kt:
```diff
@@ -464,9 +468,16 @@ class CardReaderPaymentViewModel
-        reFetchOrder()
+        if (arguments.paymentOrRefund.isPOS) {
+            launch {
+                orderRepository.fetchOrderById(orderId)
```
np: I think this worth a comment (or better to extract this to a function and call it e.g. `fetchAndCacheOrder` as the code is not self-explanatory in this case. It looks like that we fetch an order and don't use it at all. The name `fetchOrderById` doesn't help either, as it's not clear that it creates a side effect - stores the order in DB

samiuelson replied:
Done – ab16fefc7a97b0d9583507de4f679268d207d18a

kidinov reviewed (approved):
LGTM! I left one NP - feel free to ignore

kidinov committed:
Merge branch 'trunk' into 11628-return-to-pos-right-after-payment-success

samiuelson committed:
Clean up the code
